### DISCUSSION TITLE
Autovalue db_select MAX(ID)+1

### DISCUSSION THREAD

**Discussion by Andreas-Lampert** [e-netze/webgis-community#42] [General]
Hallo Jürgen!

Wir verwenden beim Editieren ein SQL-Statement zum automatischen Befüllen einer ID-Spalte mit dem nächstfolgenden Zahlen-Wert.
 
Im CMS ist im Bereich Auto Value db_select ausgewählt und beim Benutzerdefinierten Auto Value 2 ein SQLStatement der Form SELECT MAX(ID)+1 from MAX_ID_PROC.

Beim Erstellen eines neues Objekts wird die nächstfolgende ID (z.B. 9999) automatisch vergeben und in die DB geschrieben.

Ein Problem gibt es allerdings wenn das bestehende Objekt bearbeitet wird. 
Sobald dann auf Speichern geklickt wird, wird die ID wieder um eins hoch gezählt (im Beispiel dann 10000). Und das obwohl das Feld nur Readonly ist und damit natürlich keine Änderung der ID vom User vorgenommen wird. Bei jedem Mal speichern erhöht sich die ID um +1. 

Die IDs müssen unbedingt stabil bleiben, da nach der Erstellung des Objektes anhand der ID diverse Dokumente darauf gejoint werden. Eine Bearbeitung des bestehendes Objekts ist auch notwendig, da immer wieder Änderungen eingearbeitet werden. 

Wie können wir beim Erstellen eines Objekts im webgis die nächst höhere ID automatisch vergeben lassen und garantieren, dass die ID beim Bearbeiten des Objektes unter keinen Umständen mehr geändert wird?

Danke und einen schönen Dienstag!

Lg, Andi

**Comment by gis-eni** [comment]
Hallo Andi,

ich kann das Problem nachvollziehen. Grundsätzlich wird der Autovalue vom Type ``db_select`` immer angewendet, wenn ein Objekt Editiert wird. Also sowohl beim Erstellen als auch beim Ändern.

Um das Problem zu lösen, kann dir ab Release 5.23.901 einer weiteren Autovalue Type ``db_select_on_inesrt`` anbieten. Der wird dann nur noch beim INSERT (also beim Erstellen eines Objektes) angewendet.

Allerdings geb ich hier zu bedenken:
``db_select`` Autovalues sind eigentlich dazu da, dass beispielsweise Texte für die Beschriftung es Objektes angehängt werden:
* Der Anwender wähl für ein Objekt einen oder mehrere Attribute per Auswahlliste aus.
* Beim Speichern werden aus den "Codes" für diese Auswahllisten dann aus anderen Tabellen sprechende Texte geholt, damit das Objekt in der Karte ordentlich beschriftet wird.
* ...

Solche Anwendungsfälle sind nicht unbedingt kritisch. Im schlimmsten Fall wird ein Objekt falsch oder nicht beschriftet.

Wenn du hier von eindeutigen IDs sprichst, die immer konsistent bleiben sollten, würde ich das vielleicht nicht über Autovalues lösen. Du hast schon eine SQL PROC/FUNCTION, die die nächste ID berechnet. Dann würde ich in diesem Fall auch gleich in der Datenbank einen INSERT TRIGGER machen, mit dem die IDs geschrieben werden. Das ist wesentlich sicherer. Damit ist auch gewährleistet, dass die IDs gesetzt werden, wenn Objekte nicht über WebGIS erstellt werden (beispielsweise im ArcMap).
